frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx

```diff
@@ -4,21 +4,143 @@ import { Select, Modal } from 'antd'
 import { eventUsageLogic } from 'lib/utils/eventUsageLogic'
 import { saveToDashboardModalLogic } from 'lib/components/SaveToDashboard/saveToDashboardModalLogic'
 import { dashboardsModel } from '~/models/dashboardsModel'
-import { InsightModel } from '~/types'
 import { insightLogic } from 'scenes/insights/insightLogic'
 import { lemonToast } from '../lemonToast'
 import { router } from 'kea-router'
 import { urls } from 'scenes/urls'
+import './AddToDashboard.scss'
+import { IconMagnifier } from 'lib/components/icons'
+import { LemonInput } from 'lib/components/LemonInput/LemonInput'
+import { List, ListRowProps, ListRowRenderer } from 'react-virtualized/dist/es/List'
+import { AutoSizer } from 'react-virtualized/dist/es/AutoSizer'
+import { LemonButton } from 'lib/components/LemonButton'
+import { Link } from 'lib/components/Link'
+import { DashboardType, InsightModel } from '~/types'
+import clsx from 'clsx'
+import { LemonDivider } from 'lib/components/LemonDivider'
+import { LemonModal } from 'lib/components/LemonModal'
 
 interface SaveToDashboardModalProps {
     visible: boolean
     closeModal: () => void
     insight: Partial<InsightModel>
 }
 
+export function AddToDashboardModal({ visible, closeModal, insight }: SaveToDashboardModalProps): JSX.Element {
+    const logic = saveToDashboardModalLogic({
+        id: insight.short_id,
+        insight: insight,
+        fromDashboard: insight.dashboards?.[0] || undefined,
+    })
+
+    const { searchQuery, currentDashboards, orderedDashboards, scrollIndex } = useValues(logic)
+    const { setSearchQuery } = useActions(logic)
+    const { addNewDashboard } = useActions(logic)
+    const { reportSavedInsightToDashboard, reportRemovedInsightFromDashboard } = useActions(eventUsageLogic)
+    const { insightLoading } = useValues(insightLogic)
+    const { updateInsight } = useActions(insightLogic)
+
+    async function addToDashboard(dashboardToAdd: number): Promise<void> {
+        updateInsight({ ...insight, dashboards: [...(insight.dashboards || []), dashboardToAdd] }, () => {
+            reportSavedInsightToDashboard()
+            lemonToast.success('Insight added to dashboard', {
+                button: {
+                    label: 'View dashboard',
+                    action: () => router.actions.push(urls.dashboard(dashboardToAdd)),
+                },
+            })
+        })
+    }
+
+    async function removeFromDashboard(dashboardToRemove: number): Promise<void> {
+        updateInsight(
+            {
+                ...insight,
+                dashboards: (insight.dashboards || []).filter((d) => d !== dashboardToRemove),
+            },
+            () => {
+                reportRemovedInsightFromDashboard()
+                lemonToast.success('Insight removed from dashboard')
+            }
+        )
+    }
```

Comment:
Can these functions be moved to `saveToDashboardModalLogic`? Ideally we want to move as much data manipulation complexity into kea as possible.

Reply:
Hmm, that's what the logics got upset about when I tried to have the new dashboard auto selected 🤔